### PR TITLE
Broadcast block only once

### PR DESCRIPTION
**Pull request by AdoAdoAdo**

### CODE REVIEW

File: consensus/spos/interface.go [hunk 1]
```diff
@@ -116,6 +118,12 @@ type WorkerHandler interface {
 	IsInterfaceNil() bool
 }
 
+// PoolAdder adds data in a key-value pool
+type PoolAdder interface {
+	Put(key []byte, value interface{}) (evicted bool)
```
Comment (iulianpascalau) [inline consensus/spos/interface.go:123]: :+1:

File: consensus/spos/worker.go [hunk 2]
```diff
@@ -58,70 +61,62 @@ type Worker struct {
...
-	rounder consensus.Rounder,
-	shardCoordinator sharding.Coordinator,
-	singleSigner crypto.SingleSigner,
-	syncTimer ntp.SyncTimer,
-	headerSigVerifier RandSeedVerifier,
-	chainID []byte,
-	networkShardingCollector consensus.NetworkShardingCollector,
-	antifloodHandler consensus.P2PAntifloodHandler,
-) (*Worker, error) {
+func NewWorker(args *WorkerArgs) (*Worker, error) {
```
Comment (iulianpascalau) [inline consensus/spos/worker.go:91]: :+1:

Comment (iulianpascalau) [inline consensus/spos/worker.go:91]: If we pass the argument as value, the check from line 144 becomes useless (worker is instantiated once, no performance impact)

File: consensus/spos/worker.go [hunk 3]
```diff
@@ -384,8 +372,13 @@ func (wrk *Worker) ProcessReceivedMessage(message p2p.MessageP2P, fromConnectedP
 
 	go wrk.updateNetworkShardingVals(message, cnsDta)
 
+	isMessageWithBlockBody := wrk.consensusService.IsMessageWithBlockBody(msgType)
 	isMessageWithBlockHeader := wrk.consensusService.IsMessageWithBlockHeader(msgType)
 	isMessageWithBlockBodyAndHeader := wrk.consensusService.IsMessageWithBlockBodyAndHeader(msgType)
+	if isMessageWithBlockBody || isMessageWithBlockHeader {
```
Comment (iulianpascalau) [inline consensus/spos/worker.go:378]: if isMessageWithBlockBody || **isMessageWithBlockBodyAndHeader** ?

Comment (AdoAdoAdo) [reply consensus/spos/worker.go:378]: yes, that was the intention
good find

File: consensus/spos/worker.go [hunk 4]
```diff
@@ -58,70 +61,62 @@ type Worker struct {
 	mutReceivedHeadersHandler sync.RWMutex
 
 	antifloodHandler consensus.P2PAntifloodHandler
+	poolAdder        PoolAdder
+}
+
+// WorkerArgs holds the consensus worker arguments
+type WorkerArgs struct {
```
Comment (SebastianMarian) [inline consensus/spos/worker.go:68]: :+1:

**Pull request merged**